services/requester/batch_tx_pool.go

```diff
@@ -151,20 +143,46 @@ func (t *BatchTxPool) Add(
 	// For all 3 cases, we record the activity time for the next
 	// transactions that might come from the same EOA.
 	// [X] is equal to the configured `TxBatchInterval` duration.
-	lastActivityTime, found := t.eoaActivity.Get(from)
-	if !found || time.Since(lastActivityTime) > t.config.TxBatchInterval {
-		select {
-		case <-t.done:
-			return fmt.Errorf("the server is shutting down")
-		default:
-			t.txChan <- hexEncodedTx
+	lastActivity, found := t.eoaActivity.Get(from)
+	txHash := tx.Hash()
+
+	// Reject transactions that have already been submitted,
+	// as they are *bound* to fail. Two transactions with
+	// identical hashes, are expected to have the exact same
+	// payload.
+	// - If the 1st was invalid, with a nonce mismatch, the 2nd
+	// is guaranteed to be invalid as well.
+	// - If the 1st was valid, the 2nd is also guaranteed to be
+	// invalid with a nonce mismatch.
+	// In both cases, there is no reason to wasting any signing
+	// keys.
+	if found && slices.Contains(lastActivity.txHashes, txHash) {
+		return fmt.Errorf(
+			"%w: a tx with hash %s has already been submitted",
+			errs.ErrInvalid,
+			txHash,
+		)
+	}
+
+	if !found || time.Since(lastActivity.submittedAt) > t.config.TxBatchInterval {
```

Comment:
why it's `||` not `&&`? can you add comments? there are quite 3 cases here (from 2 if, and 1 else). 
```suggestion
	if !found || time.Since(lastActivity.submittedAt) > t.config.TxBatchInterval {
```

Reply:
This condition covers both cases below:

```
1. EOA activity not found:
=> We send the transaction individually, without adding it
to the batch pool.

2. EOA activity found AND it was more than [X] seconds ago:
=> We send the transaction individually, without adding it
to the batch pool.
```

The `!found` part is for case 1. The `time.Since(lastActivity.submittedAt) > t.config.TxBatchInterval` part is for case 2. So we need `||` and not `&&` to cover both cases.

Do you think it would be more readable with an `else if` instead? But the body would still be the same for both `if` and `else if`.

Reply:
yeah, it's hard to think about the else case when the `if` case have a `||` . better we make the condition a big clear, and add comments to describe.

Reply:
Updated in https://github.com/onflow/flow-evm-gateway/pull/852/commits/6a21116a5c90078fcae6303b7b65f25b502372a8 . Let me know what you think.